File: Tim/WarTicker.java
package Tim;

import java.text.DecimalFormat;
import java.util.HashSet;
import java.util.Timer;
import java.util.TimerTask;

import Tim.Data.WordWar;

public class WarTicker {
	public HashSet<WordWar> wars;
	WarClockThread warTicker;

	private WarTicker() {
		Timer ticker;

		this.wars = Tim.db.loadWars();

		this.warTicker = new WarClockThread(this);
		ticker = new Timer(true);
		ticker.scheduleAtFixedRate(this.warTicker, 0, 1000);
	}

	/**
	 * Singleton access method.
	 *
	 * @return Singleton
	 */
	public static WarTicker getInstance() {
		return SingletonHelper.instance;
	}

	private void _tick() {
		this._warsUpdate();
	}

	private void _warsUpdate() {
		if (this.wars != null && this.wars.size() > 0) {
			long currentEpoch = System.currentTimeMillis() / 1000;

			for (WordWar war : this.wars) {
				if (war.startEpoch >= currentEpoch) {
					long timeDifference = war.startEpoch - currentEpoch;

					switch ((int) timeDifference) {
						case 60:
						case 30:
						case 5:
						case 4:
						case 3:
						case 2:
						case 1:
							this.warStartCount(war);
							break;

						case 0:
							this.beginWar(war);
							break;

						default:
							if (timeDifference >= (60 * 60)) {
								// Alert hourly at 1+ hours of wait
								if (timeDifference % (60 * 60) == 0) {
									this.warStartCount(war);
								}
							} else if (timeDifference >= (60 * 30)) {
								// Alert at 30 minutes
								if (timeDifference % (60 * 30) == 0) {
									this.warStartCount(war);
								}
							} else if (timeDifference == (60 * 10)) {
								// Alert at 10 minutes
								this.warStartCount(war);
							} else if (timeDifference == (60 * 5)) {
								// Alert at 5 minutes
								this.warStartCount(war);
							}
					}
				} else {
					if (war.endEpoch >= currentEpoch) {
						if (war.warState == WordWar.State.PENDING) {
							// The war should have already started at this point, but for some reason hasn't.
							this.beginWar(war);
						} else {
							long timeDifference = war.endEpoch - currentEpoch;

							switch ((int) timeDifference) {
								case 60:
								case 5:
								case 4:
								case 3:
								case 2:
								case 1:
									this.warEndCount(war);
									break;

								case 0:
									this.endWar(war);
									break;

								default:
									if (timeDifference >= (60 * 60)) {
										if (timeDifference % (60 * 60) == 0) {
											this.warEndCount(war);
										}
									} else if (timeDifference >= (60 * 30)) {
										if (timeDifference % (60 * 30) == 0) {
											this.warEndCount(war);
										}
									} else if (timeDifference == (60 * 10)) {
										this.warEndCount(war);
									} else if (timeDifference == (60 * 5)) {
										this.warEndCount(war);
									}
							}
						}
					} else {
						this.endWar(war);
					}
				}
			}
		}
	}

	private void warStartCount(WordWar war) {
		long timeToStart = war.startEpoch - (System.currentTimeMillis() / 1000);

		if (timeToStart < 60) {
			Tim.bot.sendIRC()
				   .message(war.getChannel(), war.getSimpleName() + ": Starting in " + timeToStart + (timeToStart == 1 ? " second" : " seconds") + ".");
		} else {
			int minutes = (int) timeToStart / 60;
			if (minutes * 60 == timeToStart) {
				Tim.bot.sendIRC()
					   .message(war.getChannel(), war.getName(false, true) + ": Starting in " + minutes + (minutes == 1 ? " minute" : " minutes") + ".");
			} else {
				Tim.bot.sendIRC()
					   .message(war.getChannel(),
								war.getName(false, true) + ": Starting in " + new DecimalFormat("###.#").format(timeToStart / 60.0) + " minutes.");
			}
		}
	}

	private void beginWar(WordWar war) {
		long   currentEpoch = System.currentTimeMillis() / 1000;
		long   lateStart    = currentEpoch - war.startEpoch;
		String appendString = "";

		if (lateStart >= 5) {
			Tim.logErrorString(String.format("War starting %d seconds late in %s.", lateStart, war.channel));
			appendString = String.format(" (%d seconds late. Sorry!)", lateStart);
		}

		Tim.bot.sendIRC()
			   .message(war.getChannel(), war.getSimpleName() + ": Starting now!" + appendString);
		this.notifyWarMembers(war, war.getSimpleName() + " starts now!" + appendString);
		war.beginWar();
	}

	private void warEndCount(WordWar war) {
		long timeToEnd = war.endEpoch - (System.currentTimeMillis() / 1000);

		if (timeToEnd < 60) {
			Tim.bot.sendIRC()
				   .message(war.getChannel(), war.getSimpleName() + ": " + timeToEnd + (timeToEnd == 1 ? " second" : " seconds") + " remaining!");
		} else {
			int remaining = (int) timeToEnd / 60;
			Tim.bot.sendIRC()
				   .message(war.getChannel(), war.getSimpleName() + ": " + remaining + (remaining == 1 ? " minute" : " minutes") + " remaining.");
		}
	}

	private void endWar(WordWar war) {
		long currentEpoch = System.currentTimeMillis() / 1000;
		long lateEnd      = currentEpoch - war.endEpoch;

		String appendString = "";
		if (lateEnd >= 5) {
			Tim.logErrorString(String.format("War ending %d seconds late in %s.", lateEnd, war.channel));
			appendString = String.format(" (%d seconds late. Sorry!)", lateEnd);
		}

		Tim.bot.sendIRC()
			   .message(war.getChannel(), war.getSimpleName() + ": Ending now!" + appendString);
		this.notifyWarMembers(war, war.getSimpleName() + " is over!" + appendString);

		if (war.currentChain >= war.totalChains) {
			war.endWar();
			this.wars.remove(war);
		} else {
			war.currentChain++;

			if (war.randomness) {
				war.startEpoch = war.endEpoch + (long) ((war.baseBreak) + (war.baseBreak * ((Tim.rand.nextInt(20) - 10) / 100.0)));
				war.endEpoch = war.startEpoch + (long) (war.baseDuration + (war.baseDuration * ((Tim.rand.nextInt(20) - 10) / 100.0)));
			} else {
				war.startEpoch = war.endEpoch + war.baseBreak;
				war.endEpoch = war.startEpoch + war.baseDuration;
			}

			war.chainWarBreak();
			warStartCount(war);
		}
	}

	private void notifyWarMembers(WordWar war, String notice) {
		for (String nick : war.warMembers) {
			Tim.bot.sendIRC()
				   .message(nick, notice);
		}
	}

	private static class SingletonHelper {
		private static final WarTicker instance = new WarTicker();
	}

	class WarClockThread extends TimerTask {
		private final WarTicker parent;

		WarClockThread(WarTicker parent) {
			this.parent = parent;
		}

		@Override
		public void run() {
			try {
				this.parent._tick();
			} catch (Throwable ex) {
				Tim.printStackTrace(ex);
			}
		}
	}
}

File: timmy/core/warticker.py
import random
import time
from datetime import timedelta
from typing import Set

from timeloop import Timeloop

from timmy import core
from timmy.data.war_state import WarState
from timmy.data.word_war import WordWar
from timmy.db_access import word_war_db

war_timer = Timeloop()


class WarTicker:
    def __init__(self):
        self.loaded_wars: Set[WordWar] = set()
        self.active_wars: Set[WordWar] = set()

    def load_wars(self) -> None:
        self.loaded_wars = word_war_db.load_wars()

        war_timer.start()

    def begin_war(self, war: WordWar) -> None:
        current_epoch = time.time()
        late_start = int(current_epoch - war.start_epoch)

        message = "{}: Starting now!".format(war.get_name())

        if late_start >= 5:
            message += " ({:d} seconds late. Sorry!)".format(late_start)

        core.bot_instance.connection.privmsg(war.channel, message)

        self.notify_war_members(war, message)

        war.begin_war()

    def end_war(self, war: WordWar) -> None:
        current_epoch = time.time()
        late_end = int(current_epoch - war.end_epoch)

        message = "{}: Ending now!".format(war.get_name())

        if late_end >= 5:
            message += " ({:d} seconds late. Sorry!)".format(late_end)

        core.bot_instance.connection.privmsg(war.channel, message)

        self.notify_war_members(war, message)

        if war.channel in core.bot_instance.channels:
            core.bot_instance.channels[war.channel].last_war_id = war.get_id()

        if war.current_chain >= war.total_chains:
            war.end_war()
            self.active_wars.remove(war)

            if war.channel in core.bot_instance.channels \
                    and core.bot_instance.channels[war.channel].newest_war_id == war.get_id():
                core.bot_instance.channels[war.channel].newest_war_id = ""
        else:
            war.current_chain += 1

            if war.randomness:
                war.start_epoch = war.end_epoch + war.base_break + (war.base_break * (random.randrange(20) - 10)) / 100
                war.end_epoch = war.start_epoch + war.base_duration + (
                        war.base_duration * (random.randrange(20) - 10)) / 100

                war.duration = war.end_epoch - war.start_epoch
            else:
                war.start_epoch = war.end_epoch + war.base_break
                war.end_epoch = war.start_epoch + war.base_duration

            war.start_break()
            self.war_start_count(war)

            if war.channel in core.bot_instance.channels:
                core.bot_instance.channels[war.channel].newest_war_id = war.get_id()

    @staticmethod
    def war_start_count(war: WordWar) -> None:
        time_to_start = int(war.start_epoch - time.time())

        if time_to_start < 60:
            message = "{}: Starting in {:d} {}.".format(
                    war.get_name(), time_to_start, "seconds" if time_to_start > 1 else "second"
            )
        else:
            minutes = time_to_start / 60
            if time_to_start % 60 == 0:
                message = "{}: Starting in {:d} {}.".format(
                        war.get_name(include_duration = True), int(minutes), "minutes" if minutes > 1 else "minute"
                )
            else:
                message = "{}: Starting in {:.1f} minutes.".format(war.get_name(include_duration = True), minutes)

        core.bot_instance.connection.privmsg(war.channel, message)

    @staticmethod
    def war_end_count(war: WordWar) -> None:
        time_to_end = int(war.end_epoch - time.time())

        if time_to_end < 60:
            message = "{}: {:d} {} remaining!".format(
                    war.get_name(), time_to_end, "seconds" if time_to_end > 1 else "second"
            )
        else:
            minutes = time_to_end // 60
            message = "{}: {:d} {} remaining.".format(
                    war.get_name(), minutes, "minutes" if minutes > 1 else "minute"
            )

        core.bot_instance.connection.privmsg(war.channel, message)

    @staticmethod
    def notify_war_members(war: WordWar, message: str) -> None:
        for nick in war.war_members:
            core.bot_instance.connection.privmsg(nick, message)


@war_timer.job(interval = timedelta(seconds = 1))
def war_update_loop() -> None:
    from timmy.core import bot_instance
    loaded_wars = core.war_ticker.loaded_wars.copy()

    for war in loaded_wars:
        if war.channel in bot_instance.channels.keys():
            core.war_ticker.active_wars.add(war)
            core.war_ticker.loaded_wars.remove(war)

            bot_instance.channels[war.channel].newest_war_id = war.get_id()

    wars = core.war_ticker.active_wars.copy()
    if wars is None or len(wars) <= 0:
        return

    current_epoch = time.time()

    for war in wars:
        if war.start_epoch >= current_epoch:
            time_difference = int(war.start_epoch - current_epoch)

            if time_difference in [600, 300, 60, 30, 5, 4, 3, 2, 1]:
                core.war_ticker.war_start_count(war)
            elif time_difference == 0:
                core.war_ticker.begin_war(war)
            elif time_difference >= 3600:
                if time_difference % 3600 == 0:
                    core.war_ticker.war_start_count(war)
            elif time_difference >= 1800:
                if time_difference % 1800 == 0:
                    core.war_ticker.war_start_count(war)
        else:
            if war.end_epoch >= current_epoch:
                if war.state == WarState.PENDING:
                    core.war_ticker.begin_war(war)
                else:
                    time_difference = int(war.end_epoch - current_epoch)

                    if time_difference in [600, 300, 60, 5, 4, 3, 2, 1]:
                        core.war_ticker.war_end_count(war)
                    elif time_difference == 0:
                        core.war_ticker.end_war(war)
                    elif time_difference >= 3600:
                        if time_difference % 3600 == 0:
                            core.war_ticker.war_end_count(war)
                    elif time_difference >= 1800:
                        if time_difference % 1800 == 0:
                            core.war_ticker.war_end_count(war)
            else:
                core.war_ticker.end_war(war)
